File: Diagrammatize.Design.Pattern/code/cxx/Interpreter/A1/turtle/TurtleCanvas.cpp
// TurtleCanvas.cpp
#include "TurtleCanvas.hpp"
#include "language/ExecuteException.hpp"
#include <QDebug>

class GoExecutor;
class DirectionExecutor;

const int TurtleCanvas::UNIT_LENGTH;
const int TurtleCanvas::DIRECTION_UP;
const int TurtleCanvas::DIRECTION_RIGHT;
const int TurtleCanvas::DIRECTION_DOWN;
const int TurtleCanvas::DIRECTION_LEFT;
const int TurtleCanvas::RELATIVE_DIRECTION_RIGHT;
const int TurtleCanvas::RELATIVE_DIRECTION_LEFT;
const int TurtleCanvas::RADIUS;

// 前向声明内部类
class TurtleExecutor : public Executor
{
protected:
    TurtleCanvas* canvas_;
    
public:
    TurtleExecutor(TurtleCanvas* canvas) : canvas_(canvas) {}
    virtual ~TurtleExecutor() {}
};

class GoExecutor : public TurtleExecutor
{
public:
    GoExecutor(TurtleCanvas* canvas) : TurtleExecutor(canvas) {}
    
    void execute() override {
        canvas_->go(TurtleCanvas::UNIT_LENGTH);
    }
};

class DirectionExecutor : public TurtleExecutor
{
private:
    int relativeDirection;
    
public:
    DirectionExecutor(TurtleCanvas* canvas, int relativeDirection) 
        : TurtleExecutor(canvas), relativeDirection(relativeDirection) {}
    
    void execute() override {
        canvas_->setRelativeDirection(relativeDirection);
    }
};

// TurtleCanvas 实现
TurtleCanvas::TurtleCanvas(int width, int height, QWidget* parent)
    : QWidget(parent), direction_(0), executor_(nullptr)
{
    setFixedSize(width, height);
    initialize();
}

TurtleCanvas::~TurtleCanvas()
{
}

void TurtleCanvas::setExecutor(Executor* executor)
{
    executor_ = executor;
}

void TurtleCanvas::setRelativeDirection(int relativeDirection)
{
    setDirection(direction_ + relativeDirection);
}

void TurtleCanvas::setDirection(int direction)
{
    if (direction < 0) {
        direction = 12 - (-direction) % 12;
    } else {
        direction = direction % 12;
    }
    direction_ = direction % 12;
}

void TurtleCanvas::go(int length)
{
    int newx = position_.x();
    int newy = position_.y();
    
    switch (direction_) {
    case DIRECTION_UP:
        newy -= length;
        break;
    case DIRECTION_RIGHT:
        newx += length;
        break;
    case DIRECTION_DOWN:
        newy += length;
        break;
    case DIRECTION_LEFT:
        newx -= length;
        break;
    default:
        break;
    }
    
    QPainter painter(this);
    painter.setPen(Qt::red);
    painter.setBrush(Qt::red);
    
    // 画线
    painter.drawLine(position_.x(), position_.y(), newx, newy);
    // 画圆点
    painter.drawEllipse(QPoint(newx, newy), RADIUS, RADIUS);
    
    position_.setX(newx);
    position_.setY(newy);
}

std::unique_ptr<Executor> TurtleCanvas::createExecutor(const std::string& name)
{
    if (name == "go") {
        return std::make_unique<GoExecutor>(this);
    } else if (name == "right") {
        return std::make_unique<DirectionExecutor>(this, RELATIVE_DIRECTION_RIGHT);
    } else if (name == "left") {
        return std::make_unique<DirectionExecutor>(this, RELATIVE_DIRECTION_LEFT);
    } else {
        return nullptr;
    }
}

void TurtleCanvas::initialize()
{
    position_ = QPoint(width() / 2, height() / 2);
    direction_ = 0;
    
    QPalette pal = palette();
    pal.setColor(QPalette::Foreground, Qt::red);
    pal.setColor(QPalette::Background, Qt::white);
    setPalette(pal);
    setAutoFillBackground(true);
    
    update(); // 清空画布
}

void TurtleCanvas::paintEvent(QPaintEvent* event)
{
    Q_UNUSED(event);
    initialize();
    
    if (executor_ != nullptr) {
        try {
            executor_->execute();
        } catch (const ExecuteException& e) {
            // 处理异常，可以添加日志输出
            qDebug() << "ExecuteException caught:" << e.what();
        }
    }
}

File: Diagrammatize.Design.Pattern/code/python/Interpreter/A1/turtle/TurtleCanvas.py

from tkinter import Canvas
from language import Executor
from language import ExecutorFactory
from language import ExecuteException

class TurtleCanvas(Canvas, ExecutorFactory):
    UNIT_LENGTH = 30  # 前进时的长度单位
    DIRECTION_UP = 0  # 上方
    DIRECTION_RIGHT = 3  # 右方
    DIRECTION_DOWN = 6  # 下方
    DIRECTION_LEFT = 9  # 左方
    RELATIVE_DIRECTION_RIGHT = 3  # 右转
    RELATIVE_DIRECTION_LEFT = -3  # 左转
    RADIUS = 3  # 半径
    
    def __init__(self, master, width, height):
        super().__init__(master, width=width, height=height, bg='white')
        self.direction = 0
        self.position = None
        self.executor = None
        self.master = master
        self.width = width
        self.height = height
        self.initialize()
    
    def set_executor(self, executor):
        self.executor = executor
    
    def set_relative_direction(self, relative_direction):
        self.set_direction(self.direction + relative_direction)
    
    def set_direction(self, direction):
        if direction < 0:
            direction = 12 - (-direction) % 12
        else:
            direction = direction % 12
        self.direction = direction % 12
    
    def go(self, length):
        newx = self.position[0]
        newy = self.position[1]
        
        if self.direction == self.DIRECTION_UP:
            newy -= length
        elif self.direction == self.DIRECTION_RIGHT:
            newx += length
        elif self.direction == self.DIRECTION_DOWN:
            newy += length
        elif self.direction == self.DIRECTION_LEFT:
            newx -= length
        
        # 绘制线条
        self.create_line(self.position[0], self.position[1], newx, newy, fill='red')
        # 绘制圆点
        self.create_oval(newx - self.RADIUS, newy - self.RADIUS, 
                        newx + self.RADIUS, newy + self.RADIUS, 
                        fill='red', outline='red')
        
        self.position = (newx, newy)
    
    def createExecutor(self, name):
        if name == "go":
            return GoExecutor(self)
        elif name == "right":
            return DirectionExecutor(self, self.RELATIVE_DIRECTION_RIGHT)
        elif name == "left":
            return DirectionExecutor(self, self.RELATIVE_DIRECTION_LEFT)
        else:
            return None
    
    def initialize(self):
        self.position = (self.width // 2, self.height // 2)
        self.direction = 0
        self.delete("all")  # 清除画布
        self.configure(bg='white')
    
    def paint(self):
        """重新绘制画布"""
        self.initialize()
        if self.executor is not None:
            try:
                self.executor.execute()
            except ExecuteException:
                pass

class TurtleExecutor(Executor):
    def __init__(self, canvas):
        self.canvas = canvas

class GoExecutor(TurtleExecutor):
    def __init__(self, canvas):
        super().__init__(canvas)
    
    def execute(self):
        self.canvas.go(TurtleCanvas.UNIT_LENGTH)

class DirectionExecutor(TurtleExecutor):
    def __init__(self, canvas, relative_direction):
        super().__init__(canvas)
        self.relative_direction = relative_direction
    
    def execute(self):
        self.canvas.set_relative_direction(self.relative_direction)

